M+ History

billriess opened the pull request:
Added M+ History for the last 4 runs.

![image](https://github.com/user-attachments/assets/78fe848a-f802-4d51-b5dd-e54fad5a0903)

Improves #159 

<!-- If needed, link to design -->

# Description 
This adds the last 4 runs to the Game Menu (ESC) screen.  We are unable to pull (to my knowledge) the duration of the key from the history API so the only way to show timers would be to rewrite using raider.io helpers and adding the optional dependency or we could record at the end of the run on the `CHALLENGE_MODE_COMPLETED `  event and store it.

# To test

- [ ] Hit ESC with keys done.

billriess committed:
Update GameMenuButton.lua

Added M+ History for the last 4 runs.

billriess committed:
Update GameMenuButton.lua

Fix sort bug.

billriess committed:
Update GameMenuButton.lua

Update history to show failed runs and color runs by success or fail

billriess committed:
Update Profile.lua

Initialize the M+ history limit

billriess committed:
Update GameMenuButton.lua

Dynamically create history list based on limit

billriess committed:
Update GameMenuSkin.lua

Add new history limit slider

billriess commented:
Made the history limit configurable
![image](https://github.com/user-attachments/assets/aa557143-45f8-47a6-9782-abb589041c56)

billriess committed:
Update GameMenuButton.lua

Minor cleanup

Toxicom commented on Core/Profile.lua:
```diff
@@ -258,6 +258,7 @@ P.addons = {
     showTips = true,
     showCollections = TXUI.IsRetail,
     showMythic = TXUI.IsRetail,
+    historyLimit = 4,
```
```suggestion
    mythicHistoryLimit = 4,
```

Toxicom commented on Modules/Misc/GameMenuButton.lua:
```diff
@@ -179,22 +197,50 @@ function M:GameMenuButton()
+          if historyFrame then
+            local historyRun = history[#history - i + 1]
+            local historyText
+            local historyTextPrefix = i .. ": "
```
```suggestion
            local historyTextPrefix = i .. ". "
```

Toxicom commented on Modules/Options/Skins/GameMenuSkin.lua:
```diff
@@ -154,6 +157,24 @@ function O:Skins_ElvUI()
+      set = function(_, value)
+        E.db.TXUI.addons.gameMenuSkin.historyLimit = value
+      end,
+      disabled = optionsDisabled and not E.db.TXUI.addons.gameMenuSkin.showMythic,
```
Does this actually become disabled when you disable `showMythic`?

billriess replied:
It did in my test. I’ll test again tonight to be certain.

Toxicom commented on Modules/Options/Skins/GameMenuSkin.lua:
```diff
@@ -154,6 +157,24 @@ function O:Skins_ElvUI()
+      desc = "Number of Mythic+ dungeons shown in the history list.",
+      min = 1,
+      max = 10,
+      step = 1,
```
How would the user disable displaying the history? It either needs a separate toggle or allow it to go to 0 and update code to reflect that. From an UX standpoint, separate toggle would make more sense.

billriess replied:
I guess I assumed we wouldn’t control just history since the m+ section would only contain current key at that point. You could argue to not show current key too so I just left it like this. If you want to disable history and leave current key on then I can add that let me know.

Toxicom reviewed (changes requested):
Haven't tested in game yet, but some things I've noticed that need adjusting

Toxicom commented on Modules/Misc/GameMenuButton.lua:
```diff
@@ -179,22 +197,50 @@ function M:GameMenuButton()
+            if historyRun then
+              local historyDungeonName = C_ChallengeMode.GetMapUIInfo(historyRun.mapChallengeModeID)
+              historyText = historyRun.completed and
+                  F.String.Good(("%s (+%d)"):format(historyDungeonName, historyRun.level)) or
```
Since `("%s (+%d)"):format(historyDungeonName, historyRun.level)` is identical for both cases, can we extract it?

billriess replied:
Yes, this was overlooked on a refractor.

billriess committed:
Update Core/Profile.lua

rename key from historyLimit to mythicHistoryLimit

Co-authored-by: Tomas Leonavičius <69549795+Toxicom@users.noreply.github.com>

billriess committed:
Update GameMenuSkin.lua

Added callback hook to refresh config when checkbox changes using AceConfigRegistry

billriess committed:
Update GameMenuButton.lua

Slight refactor

billriess commented:
Pushed requested changes. Used `AceConfigRegistry-3.0` to dynamically toggle the slider when `Show Mythic+` is changed.

I'm still not sure it makes sense to fully disable the history if you're enabling the Mythic+ section. If you feel strongly about it then it can easily be added.

Toxicom commented:
Gonna merge this in so I can make changes locally and finish this up

Toxicom committed:
Merge branch 'development' into development
